paddle/go/pserver/service.go

```diff
@@ -135,7 +138,10 @@ func (s *Service) SendGrads(grads []Gradient, dummy *int) error {
 	errCh := make(chan error, count)
 	for _, g := range grads {
 		go func(p Parameter, g Gradient) {
-			err := s.opt.UpdateParameter(p, g)
+			opt, err := s.optimizerMap[p.Name]
+			if err != nil {
+				err := opt.UpdateParameter(p, g)
```

Comment:
do **not** use `err:=` here, otherwise we created a new err variable, shadowing the err outside.
see: http://blog.charmes.net/2015/06/scope-and-shadowing-in-go.html
you can use tools to check for shadow: https://github.com/alecthomas/gometalinter
the tool has a emacs package, install: https://github.com/helinwang/go-emacs/blob/master/init.el#L20